flow/src/org/labkey/flow/data/FlowAssayProvider.java

```diff
@@ -500,27 +504,33 @@ public DataExchangeHandler createDataExchangeHandler()
     }
 
     @Override
-    public Container getAssociatedStudyContainer(ExpProtocol protocol, Object dataId)
+    public Set<Container> getAssociatedStudyContainers(ExpProtocol protocol, Collection<Integer> rowIds)
     {
-        ExpData data = getDataForDataRow(dataId, protocol);
-        if (data == null)
-            return null;
-        ExpRun run = data.getRun();
-        if (run == null)
-            return null;
-
-        Map<String, Object> properties = OntologyManager.getProperties(run.getContainer(), run.getLSID());
-        String targetStudyId = (String) properties.get(FlowProperty.TargetStudy.getPropertyDescriptor().getPropertyURI());
-
-        // Issue 24990: Copy-To-Study dropdown for flow resets when re-accessing feature
-        // If no target study explicitly set for this run, find the most recently used target study.
-        // The publishChooseStudy.jsp will select the study by default but still allow the user to override this selection.
-        if (targetStudyId == null)
-            targetStudyId = FlowRun.findMostRecentTargetStudy(run.getContainer());
+        Set<Container> result = new HashSet<>();
+        for (Integer dataRowId : rowIds)
+        {
+            Container container = null;
+            ExpData data = getDataForDataRow(dataRowId, protocol);
+            if (data != null)
+            {
+                ExpRun run = data.getRun();
+                if (run != null)
+                {
+                    Map<String, Object> properties = OntologyManager.getProperties(run.getContainer(), run.getLSID());
```

Comment:
Maybe cache the targetStudyId using the run rowId as the key to avoid getting the run object and getting the ontology properties for it for each of the result rows.

Reply:
Happily, OntologyManager.getProperties() does its own caching so I don't think an extra layer here is necessary.